File: iOS/ssabree/ssabree/App/AppContainer.swift
import Foundation

// MARK: - App Container Protocol

protocol AppContainer {
    var authDataStore: AuthDataStore { get }
    var authRepository: AuthRepository { get }
    var homeRepository: HomeRepository { get }
    var boardRepository: BoardRepository { get }
    var campusRepository: CampusRepository { get }
    var keywordRepository: KeywordRepository { get }
    var postRepository: PostRepository { get }
    var commentRepository: CommentRepository { get }
    var uploadRepository: UploadRepository { get }
    var myPageRepository: MyPageRepository { get }
    var groupRepository: GroupRepository { get }
    var groupService: GroupService { get }
    var messageRepository: MessageRepository { get }
    var chatRepository: ChatRepository { get }
    var myGroupRepository: MyGroupRepository { get }
    var ddayRepository: DdayRepository { get }
    var notificationRepository: NotificationRepository { get }
    var portfolioRepository: PortfolioRepository { get }
    var projectRepository: ProjectRepository { get }
    var reportRepository: ReportRepository { get }
    var inquiryRepository: InquiryRepository { get }
    var stackRepository: StackRepository { get }
    var imageRepository: ImageRepository { get }

    // MARK: - Cached ViewModels (탭에서 사용하는 ViewModel들 캐시)
    var homeViewModel: HomeViewModel { get }
    var boardViewModel: BoardViewModel { get }
    var groupViewModel: GroupViewModel { get }
    var messageViewModel: MessageViewModel { get }
    var myGroupStudyViewModel: MyGroupAllViewModel { get }
    var myGroupProjectViewModel: MyGroupAllViewModel { get }
}

// MARK: - Real App Container

final class RealAppContainer: AppContainer {
    // MARK: - Storage
    let authDataStore: AuthDataStore

    // MARK: - Services
    let groupService: GroupService

    // MARK: - Repositories
    let authRepository: AuthRepository
    let homeRepository: HomeRepository
    let boardRepository: BoardRepository
    let campusRepository: CampusRepository
    let keywordRepository: KeywordRepository
    let postRepository: PostRepository
    let commentRepository: CommentRepository
    let uploadRepository: UploadRepository
    let myPageRepository: MyPageRepository
    let groupRepository: GroupRepository
    let messageRepository: MessageRepository
    let chatRepository: ChatRepository
    let myGroupRepository: MyGroupRepository
    let ddayRepository: DdayRepository
    let notificationRepository: NotificationRepository
    let portfolioRepository: PortfolioRepository
    let projectRepository: ProjectRepository
    let reportRepository: ReportRepository
    let inquiryRepository: InquiryRepository
    let stackRepository: StackRepository
    let imageRepository: ImageRepository

    // MARK: - Cached ViewModels
    lazy var homeViewModel: HomeViewModel = HomeViewModel(homeRepository: homeRepository, myPageRepository: myPageRepository)
    lazy var boardViewModel: BoardViewModel = BoardViewModel(boardRepository: boardRepository, postRepository: postRepository)
    lazy var groupViewModel: GroupViewModel = GroupViewModel(
        groupRepository: groupRepository,
        keywordRepository: keywordRepository,
        myPageRepository: myPageRepository,
        campusRepository: campusRepository
    )
    lazy var messageViewModel: MessageViewModel = MessageViewModel(chatRepository: chatRepository)
    lazy var myGroupStudyViewModel: MyGroupAllViewModel = MyGroupAllViewModel(groupRepository: groupRepository, groupKind: .study, authRepository: authRepository)
    lazy var myGroupProjectViewModel: MyGroupAllViewModel = MyGroupAllViewModel(groupRepository: groupRepository, groupKind: .project, authRepository: authRepository)

    init() {
        // Setup secure storage with Keychain
        let secureStore = KeychainSecureStore()
        let dataStore = AuthDataStore(secureStore: secureStore)
        self.authDataStore = dataStore

        // Configure API Client with auth data store
        APIClient.shared.configure(authDataStore: dataStore)

        // Configure WebSocket Manager with auth data store
        Task { @MainActor in
            ChatWebSocketManager.shared.configure(authDataStore: dataStore)
        }

        // Initialize services
        let authService = AuthServiceImpl()
        let boardService = BoardServiceImpl()
        let campusService = CampusServiceImpl()
        let commentService = CommentServiceImpl()
        let homeService = HomeServiceImpl()
        let postService = PostServiceImpl()
        let uploadService = UploadServiceImpl()
        let myPageService = MyPageServiceImpl()
        let groupServiceImpl = GroupServiceImpl()
        self.groupService = groupServiceImpl
        let messageService = MessageServiceImpl()
        let chatService = ChatServiceImpl()
        let ddayService = DdayServiceImpl()
        let notificationService = NotificationServiceImpl()
        let portfolioService = PortfolioServiceImpl()
        let projectService = ProjectServiceImpl()
        let reportService = ReportServiceImpl()
        let inquiryService = InquiryServiceImpl()
        let stackService = StackServiceImpl()

        // Initialize repositories
        self.authRepository = AuthRepositoryImpl(
            authService: authService,
            authDataStore: dataStore
        )
        // DdayRepository 먼저 초기화 (HomeRepository에서 사용)
        self.ddayRepository = DdayRepositoryImpl(ddayService: ddayService)
        self.homeRepository = HomeRepositoryImpl(homeService: homeService, ddayRepository: ddayRepository)
        self.boardRepository = BoardRepositoryImpl(boardService: boardService)
        self.campusRepository = CampusRepositoryImpl(campusService: campusService)
        self.keywordRepository = KeywordRepositoryImpl()
        self.postRepository = PostRepositoryImpl(postService: postService, authDataStore: dataStore)
        self.commentRepository = CommentRepositoryImpl(commentService: commentService)
        self.uploadRepository = UploadRepositoryImpl(uploadService: uploadService)
        self.myPageRepository = MyPageRepositoryImpl(myPageService: myPageService)
        self.groupRepository = GroupRepositoryImpl(groupService: groupServiceImpl)
        self.messageRepository = MessageRepositoryImpl(messageService: messageService)
        self.chatRepository = ChatRepositoryImpl(chatService: chatService)
        self.myGroupRepository = MyGroupRepositoryImpl(groupService: groupServiceImpl)
        self.notificationRepository = NotificationRepositoryImpl(notificationService: notificationService)
        self.portfolioRepository = PortfolioRepositoryImpl(portfolioService: portfolioService)
        self.projectRepository = ProjectRepositoryImpl(projectService: projectService)
        self.reportRepository = ReportRepositoryImpl(reportService: reportService)
        self.inquiryRepository = InquiryRepositoryImpl(inquiryService: inquiryService)
        self.stackRepository = StackRepositoryImpl(stackService: stackService)
        self.imageRepository = ImageRepositoryImpl.shared
    }
}

// MARK: - Fake App Container (for previews and testing)

final class FakeAppContainer: AppContainer {
    let authDataStore: AuthDataStore
    let authRepository: AuthRepository
    let homeRepository: HomeRepository
    let boardRepository: BoardRepository
    let campusRepository: CampusRepository
    let keywordRepository: KeywordRepository
    let postRepository: PostRepository
    let commentRepository: CommentRepository
    let uploadRepository: UploadRepository
    let myPageRepository: MyPageRepository
    let groupRepository: GroupRepository
    let groupService: GroupService
    let messageRepository: MessageRepository
    let chatRepository: ChatRepository
    let myGroupRepository: MyGroupRepository
    let ddayRepository: DdayRepository
    let notificationRepository: NotificationRepository
    let portfolioRepository: PortfolioRepository
    let projectRepository: ProjectRepository
    let reportRepository: ReportRepository
    let inquiryRepository: InquiryRepository
    let stackRepository: StackRepository
    let imageRepository: ImageRepository

    // MARK: - Cached ViewModels
    lazy var homeViewModel: HomeViewModel = HomeViewModel(homeRepository: homeRepository, myPageRepository: myPageRepository)
    lazy var boardViewModel: BoardViewModel = BoardViewModel(boardRepository: boardRepository, postRepository: postRepository)
    lazy var groupViewModel: GroupViewModel = GroupViewModel(
        groupRepository: groupRepository,
        keywordRepository: keywordRepository,
        myPageRepository: myPageRepository,
        campusRepository: campusRepository
    )
    lazy var messageViewModel: MessageViewModel = MessageViewModel(chatRepository: chatRepository)
    lazy var myGroupStudyViewModel: MyGroupAllViewModel = MyGroupAllViewModel(groupRepository: groupRepository, groupKind: .study, authRepository: authRepository)
    lazy var myGroupProjectViewModel: MyGroupAllViewModel = MyGroupAllViewModel(groupRepository: groupRepository, groupKind: .project, authRepository: authRepository)

    init() {
        let secureStore = InMemorySecureStore()
        self.authDataStore = AuthDataStore(secureStore: secureStore)

        self.authRepository = FakeAuthRepository()
        self.homeRepository = FakeHomeRepository()
        self.boardRepository = FakeBoardRepository()
        self.campusRepository = FakeCampusRepository()
        self.keywordRepository = FakeKeywordRepository()
        self.postRepository = FakePostRepository()
        self.commentRepository = FakeCommentRepository()
        self.uploadRepository = FakeUploadRepository()
        self.myPageRepository = FakeMyPageRepository()
        self.groupRepository = FakeGroupRepository()
        self.groupService = FakeGroupService()
        self.messageRepository = FakeMessageRepository()
        self.chatRepository = FakeChatRepository()
        self.myGroupRepository = FakeMyGroupRepository()
        self.ddayRepository = FakeDdayRepository()
        self.notificationRepository = FakeNotificationRepository()
        self.portfolioRepository = FakePortfolioRepository()
        self.projectRepository = FakeProjectRepository()
        self.reportRepository = FakeReportRepository()
        self.inquiryRepository = FakeInquiryRepository()
        self.stackRepository = FakeStackRepository()
        self.imageRepository = FakeImageRepository()
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/app/AppContainer.kt
package com.ssafy.ssabree.app

import com.ssafy.ssabree.core.datasource.local.DdayLocalStore
import com.ssafy.ssabree.core.repository.AuthRepository
import com.ssafy.ssabree.core.repository.AuthRepositoryImpl
import com.ssafy.ssabree.core.repository.CampusRepository
import com.ssafy.ssabree.core.repository.CampusRepositoryImpl
import com.ssafy.ssabree.core.repository.CommentRepository
import com.ssafy.ssabree.core.repository.CommentRepositoryImpl
import com.ssafy.ssabree.core.repository.DdayRepository
import com.ssafy.ssabree.core.repository.DdayRepositoryImpl
import com.ssafy.ssabree.core.repository.GroupRepository
import com.ssafy.ssabree.core.repository.GroupRepositoryImpl
import com.ssafy.ssabree.core.repository.MyGroupRepository
import com.ssafy.ssabree.core.repository.MyGroupRepositoryImpl
import com.ssafy.ssabree.core.repository.HomeRepository
import com.ssafy.ssabree.core.repository.HomeRepositoryImpl
import com.ssafy.ssabree.core.repository.BoardRepository
import com.ssafy.ssabree.core.repository.BoardRepositoryImpl
import com.ssafy.ssabree.core.repository.UploadRepository
import com.ssafy.ssabree.core.repository.UploadRepositoryImpl
import com.ssafy.ssabree.core.repository.KeywordRepository
import com.ssafy.ssabree.core.repository.KeywordRepositoryImpl
import com.ssafy.ssabree.core.repository.InquiryRepository
import com.ssafy.ssabree.core.repository.InquiryRepositoryImpl
import com.ssafy.ssabree.core.repository.MemberRepository
import com.ssafy.ssabree.core.repository.MemberRepositoryImpl
import com.ssafy.ssabree.core.repository.PostRepository
import com.ssafy.ssabree.core.repository.PostRepositoryImpl
import com.ssafy.ssabree.core.repository.PortfolioRepository
import com.ssafy.ssabree.core.repository.PortfolioRepositoryImpl
import com.ssafy.ssabree.core.repository.ProjectRepository
import com.ssafy.ssabree.core.repository.ProjectRepositoryImpl
import com.ssafy.ssabree.core.repository.ChatRepository
import com.ssafy.ssabree.core.repository.ChatRepositoryImpl
import com.ssafy.ssabree.core.repository.ReportRepository
import com.ssafy.ssabree.core.repository.ReportRepositoryImpl
import com.ssafy.ssabree.core.repository.StackRepository
import com.ssafy.ssabree.core.repository.StackRepositoryImpl
import com.ssafy.ssabree.core.repository.test.FakeAuthRepository
import com.ssafy.ssabree.core.repository.test.FakeCampusRepository
import com.ssafy.ssabree.core.repository.test.FakeCommentRepository
import com.ssafy.ssabree.core.repository.test.FakeGroupRepository
import com.ssafy.ssabree.core.repository.test.FakeInquiryRepository
import com.ssafy.ssabree.core.repository.test.FakeMemberRepository
import com.ssafy.ssabree.core.repository.test.FakeMyGroupRepository
import com.ssafy.ssabree.core.repository.test.FakeHomeRepository
import com.ssafy.ssabree.core.repository.test.FakeBoardRepository
import com.ssafy.ssabree.core.repository.test.FakeDdayRepository
import com.ssafy.ssabree.core.repository.test.FakeUploadRepository
import com.ssafy.ssabree.core.repository.test.FakePostRepository
import com.ssafy.ssabree.core.repository.test.FakeKeywordRepository
import com.ssafy.ssabree.core.repository.test.FakePortfolioRepository
import com.ssafy.ssabree.core.repository.test.FakeProjectRepository
import com.ssafy.ssabree.core.repository.test.FakeChatRepository
import com.ssafy.ssabree.core.repository.test.FakeReportRepository
import com.ssafy.ssabree.core.repository.test.FakeStackRepository
import com.ssafy.ssabree.core.repository.NotificationRepository
import com.ssafy.ssabree.core.repository.NotificationRepositoryImpl
import com.ssafy.ssabree.core.repository.test.FakeNotificationRepository
import com.ssafy.ssabree.core.repository.ImageRepository
import com.ssafy.ssabree.core.repository.ImageRepositoryImpl
import com.ssafy.ssabree.core.repository.test.FakeImageRepository
import com.ssafy.ssabree.core.datasource.local.image.ImageCacheDatabase
import com.ssafy.ssabree.core.utils.RetrofitClient

interface AppContainer {
    val authRepository: AuthRepository
    val campusRepository: CampusRepository
    val commentRepository: CommentRepository
    val ddayRepository: DdayRepository
    val groupRepository: GroupRepository
    val myGroupRepository: MyGroupRepository
    val homeRepository: HomeRepository
    val boardRepository: BoardRepository
    val uploadRepository: UploadRepository
    val postRepository: PostRepository
    val keywordRepository: KeywordRepository
    val memberRepository: MemberRepository
    val inquiryRepository: InquiryRepository
    val portfolioRepository: PortfolioRepository
    val projectRepository: ProjectRepository
    val chatRepository: ChatRepository
    val reportRepository: ReportRepository
    val stackRepository: StackRepository
    val notificationRepository: NotificationRepository
    val imageRepository: ImageRepository
}

class RealAppContainer : AppContainer {
    private val ddayLocalStore = DdayLocalStore(ApplicationClass.appContext)

    override val authRepository: AuthRepository = AuthRepositoryImpl()
    override val campusRepository: CampusRepository = CampusRepositoryImpl()
    override val commentRepository: CommentRepository = CommentRepositoryImpl()
    override val ddayRepository: DdayRepository = DdayRepositoryImpl(ddayLocalStore)
    override val groupRepository: GroupRepository = GroupRepositoryImpl()
    override val myGroupRepository: MyGroupRepository = MyGroupRepositoryImpl()
    override val homeRepository: HomeRepository = HomeRepositoryImpl(ddayRepository)
    override val boardRepository: BoardRepository = BoardRepositoryImpl()
    override val uploadRepository: UploadRepository = UploadRepositoryImpl()
    override val postRepository: PostRepository = PostRepositoryImpl()
    override val keywordRepository: KeywordRepository = KeywordRepositoryImpl()
    override val memberRepository: MemberRepository = MemberRepositoryImpl()
    override val inquiryRepository: InquiryRepository = InquiryRepositoryImpl()
    override val portfolioRepository: PortfolioRepository = PortfolioRepositoryImpl()
    override val projectRepository: ProjectRepository = ProjectRepositoryImpl()
    override val chatRepository: ChatRepository = ChatRepositoryImpl()
    override val reportRepository: ReportRepository = ReportRepositoryImpl()
    override val stackRepository: StackRepository = StackRepositoryImpl()
    override val notificationRepository: NotificationRepository = NotificationRepositoryImpl()
    private val imageCacheDb = ImageCacheDatabase.getInstance(ApplicationClass.appContext)
    override val imageRepository: ImageRepository =
        ImageRepositoryImpl(
            dao = imageCacheDb.imageCacheDao(),
            client = RetrofitClient.client.newBuilder()
                .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                .build()
        )
}

class FakeAppContainer : AppContainer {
    override val authRepository: AuthRepository = FakeAuthRepository()
    override val campusRepository: CampusRepository = FakeCampusRepository()
    override val commentRepository: CommentRepository = FakeCommentRepository()
    override val ddayRepository: DdayRepository = FakeDdayRepository()
    override val groupRepository: GroupRepository = FakeGroupRepository()
    override val myGroupRepository: MyGroupRepository = FakeMyGroupRepository()
    override val homeRepository: HomeRepository = FakeHomeRepository()
    override val boardRepository: BoardRepository = FakeBoardRepository()
    override val uploadRepository: UploadRepository = FakeUploadRepository()
    override val postRepository: PostRepository = FakePostRepository()
    override val keywordRepository: KeywordRepository = FakeKeywordRepository()
    override val memberRepository: MemberRepository = FakeMemberRepository()
    override val inquiryRepository: InquiryRepository = FakeInquiryRepository()
    override val portfolioRepository: PortfolioRepository = FakePortfolioRepository()
    override val projectRepository: ProjectRepository = FakeProjectRepository()
    override val chatRepository: ChatRepository = FakeChatRepository()
    override val reportRepository: ReportRepository = FakeReportRepository()
    override val stackRepository: StackRepository = FakeStackRepository()
    override val notificationRepository: NotificationRepository = FakeNotificationRepository()
    override val imageRepository: ImageRepository = FakeImageRepository()
}
